app/src/main/java/org/mozilla/fenix/home/blocklist/BlocklistMiddleware.kt

```diff
@@ -0,0 +1,170 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+package org.mozilla.fenix.home.blocklist
+
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.flow.launchIn
+import kotlinx.coroutines.flow.map
+import mozilla.components.lib.state.Middleware
+import mozilla.components.lib.state.MiddlewareContext
+import mozilla.components.lib.state.Store
+import org.mozilla.fenix.home.HomeFragmentAction
+import org.mozilla.fenix.home.HomeFragmentState
+import org.mozilla.fenix.home.recentbookmarks.RecentBookmark
+import org.mozilla.fenix.home.recenttabs.RecentTab
+import org.mozilla.fenix.home.recentvisits.RecentlyVisitedItem
+
+class BlocklistMiddleware(
+    coroutineScope: CoroutineScope,
+    private val provider: HomescreenBlocklistProvider
+) : Middleware<HomeFragmentState, HomeFragmentAction> {
+
+    private var blocklist: Set<String> = setOf()
+    private var blocklistLoaded = false
+    private var store: Store<HomeFragmentState, HomeFragmentAction>? = null
+
+    init {
+        provider.blocklist.map { updatedList ->
+            blocklist = updatedList
+            if (!blocklistLoaded) {
+                filterLatestState()
+            }
+            blocklistLoaded = true
+        }.launchIn(coroutineScope)
+    }
```

Comment:
I think we should try to init this lazily when need it the first time we access it. As we be blocking when creating this middleware.

Reply:
- This data should be needed as soon as the object is constructed, in order to immediately filter the initial state of the home screen
- The `launchIn` function is a terminal flow operator that starts collection in a newly launched coroutine, so it shouldn't be blocking.